numba/cuda/reduction.py

```diff
@@ -133,37 +163,15 @@ def _type_and_size(self, dary, size):
 
         return nbtype, size
 
-    def device_partial_inplace(self, darr, size=None, init=0, stream=0):
-        """Partially reduce a device array inplace as much as possible in an
-        efficient manner. Does not automatically transfer host array.
-
-        :param darr: Used to input and output.
-        :type darr: device array
-        :param size: Number of element in ``arr``.  If None, the entire array is used.
-        :type size: int or None
-        :param init: Initial value for the reduction
-        :type init: dtype of darr
-        :param stream: All CUDA operations are performed on this stream if it is given.
-          Otherwise, a new stream is created.
-        :type stream: cuda stream
-        :returns: int -- Number of elements in ``darr`` that contains the reduction result.
-
-        """
-        if stream == 0:
-            from numba import cuda
-            stream = cuda.stream()
-            ret = self._partial_inplace_driver(darr, size, init, stream)
-            stream.synchronize()
-        else:
-            ret = self._partial_inplace_driver(darr, size, init, stream)
-        return ret
-
-    def _partial_inplace_driver(self, dary, size, init, stream):
+    def _execute_reduction(self, dary, size, stream, init=0):
 
         from numba import cuda
 
         nbtype, size = self._type_and_size(dary, size)
 
+        # Partial reduction kernels write their output strided - this kernel is
+        # used to copy the strided output to a compacted array which can be used
+        # as input to the next kernel.
```

Comment:
As a side note, do you know why a shared variable (`sm`) is used below instead of a local variable? It doesn't seem to be sharing anything useful, and it requires a syncthreads() call.